extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java

```diff
@@ -105,24 +107,30 @@
 
         Uni<JsonObject> userInfo = getUserInfoUni(vertxContext, request, resolvedContext);
 
-        return userInfo.onItem().transformToUni(
-                new Function<JsonObject, Uni<? extends SecurityIdentity>>() {
+        return userInfo.onItemOrFailure().transformToUni(
+                new BiFunction<JsonObject, Throwable, Uni<? extends SecurityIdentity>>() {
                     @Override
-                    public Uni<SecurityIdentity> apply(JsonObject userInfo) {
+                    public Uni<SecurityIdentity> apply(JsonObject userInfo, Throwable t) {
+                        if (t != null) {
+                            return Uni.createFrom().failure(new AuthenticationFailedException(t));
+                        }
                         return createSecurityIdentityWithOidcServer(vertxContext, request, resolvedContext, userInfo);
                     }
                 });
     }
 
     private Uni<SecurityIdentity> createSecurityIdentityWithOidcServer(RoutingContext vertxContext,
             TokenAuthenticationRequest request, TenantConfigContext resolvedContext, final JsonObject userInfo) {
-        Uni<TokenVerificationResult> codeFlowTokenUni = verifyTokenUni(resolvedContext,
+        Uni<TokenVerificationResult> tokenUni = verifyTokenUni(resolvedContext,
                 request.getToken().getToken());
 
-        return codeFlowTokenUni.onItem()
-                .transformToUni(new Function<TokenVerificationResult, Uni<? extends SecurityIdentity>>() {
+        return tokenUni.onItemOrFailure()
+                .transformToUni(new BiFunction<TokenVerificationResult, Throwable, Uni<? extends SecurityIdentity>>() {
                     @Override
-                    public Uni<SecurityIdentity> apply(TokenVerificationResult result) {
+                    public Uni<SecurityIdentity> apply(TokenVerificationResult result, Throwable t) {
+                        if (t != null) {
+                            return Uni.createFrom().failure(new AuthenticationFailedException(t));
```

Comment:
@cescoffier Yeah, I think I've tried it before and retried, the problem with the direct throw is that it causes `500` as opposed to `401` etc, I think `HttpSecurityRecorder` may need to be tweaked a bit to ensure this exception is still recognized, but I'd rather follow up on it later on, otherwise I suspect I'd have hard time rebasing this PR against the Vert.x 4 branch :-)